concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/window/ImmutablePercentileSampleWindow.java

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.concurrency.limits.limit.window;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+class ImmutablePercentileSampleWindow implements SampleWindow {
+    private final long minRtt;
+    private final int maxInFlight;
+    private final boolean didDrop;
+    private final List<Long> observedRtts;
+    private final double percentile;
+
+    ImmutablePercentileSampleWindow(double percentile) {
+        this.minRtt = Long.MAX_VALUE;
+        this.maxInFlight = 0;
+        this.didDrop = false;
+        this.observedRtts = new ArrayList<>();
+        this.percentile = percentile;
+    }
+
+    ImmutablePercentileSampleWindow(
+            long minRtt,
+            int maxInFlight,
+            boolean didDrop,
+            List<Long> observedRtts,
+            double percentile
+    ) {
+        this.minRtt = minRtt;
+        this.maxInFlight = maxInFlight;
+        this.didDrop = didDrop;
+        this.observedRtts = observedRtts;
+        this.percentile = percentile;
+    }
+
+    @Override
+    public ImmutablePercentileSampleWindow addSample(long rtt, int inflight) {
+        // TODO: very naive
+        // full copy in order to fulfill side-effect-free requirement of AtomicReference::updateAndGet
+        List<Long> newObservedRtts = new ArrayList<>(observedRtts);
```

Comment:
@elandau I forgot that the inner array representation of PriorityQueue is not a sorted array but somewhat trickier, so I replaced the backing collection with ArrayList. PriorityQueue was an overkill here anyway. Also covered that with some tests.

Reply:
I think this is fine for now as we experiment but this will ultimately be very inefficient since a new array is allocated as each sample is added.  Instead, we may want to just create a fixed size array and just allow for a max number of samples (maybe, 100) within a sampling window.

Reply:
> want to just create a fixed size array

hmmm, I think this won't work with existing `AtomicReference::updateAndGet`. It will require changes in the WindowedLimit architecture. Otherwise on a high congestion it might insert multiple duplicated rtts, right?

Reply:
And yes, I had plans on improving the efficiency in the next iteration.